ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala

```diff
@@ -0,0 +1,309 @@
+
+    for {
+      _ <- keyCache.putAsync(key, currentOffset, eventualValue)
+      value <- eventualValue
+    } yield keyStateToResponse(value, readers)
+  }
+
+  private def nonEmptyIntersection[T](one: Set[T], other: Set[T]): Boolean =
+    one.intersect(other).nonEmpty
+
+  private def updateOffsets(event: ContractStateEvent): Unit = {
+    cacheOffset.set(event.eventSequentialId)
+    metrics.daml.indexer.currentStateCacheSequentialIdGauge.updateValue(event.eventSequentialId)
+    event match {
+      case LedgerEndMarker(eventOffset, _) => signalNewLedgerHead(eventOffset)
+      case _ => ()
+    }
+  }
+
+  private val updateCaches: ContractStateEvent => Unit = {
+    case ContractStateEvent.Created(
+          contractId,
+          contract,
+          globalKey,
+          createLedgerEffectiveTime,
+          flatEventWitnesses,
+          _,
+          eventSequentialId,
+        ) =>
+      globalKey.foreach(
+        keyCache.put(_, eventSequentialId, Assigned(contractId, flatEventWitnesses))
+      )
+      contractsCache.put(
+        contractId,
+        eventSequentialId,
+        Active(contract, flatEventWitnesses, createLedgerEffectiveTime),
+      )
+    case ContractStateEvent.Archived(
+          contractId,
+          globalKey,
+          stakeholders,
+          _,
+          eventSequentialId,
+        ) =>
+      globalKey.foreach(keyCache.put(_, eventSequentialId, Unassigned))
+      contractsCache.put(
+        contractId,
+        eventSequentialId,
+        Archived(stakeholders),
+      )
+    case _: LedgerEndMarker => ()
+  }
+
+  private def debugEvents(
+      event: ContractStateEvent
+  )(implicit loggingContext: LoggingContext): Unit =
+    event match {
+      case ContractStateEvent.Created(
+            contractId,
+            _,
+            globalKey,
+            _,
+            _,
+            eventOffset,
+            eventSequentialId,
+          ) =>
+        logger.debug(
+          s"State events update: Created(contractId=$contractId, globalKey=$globalKey, offset=$eventOffset, eventSequentialId=$eventSequentialId)"
+        )
+      case ContractStateEvent.Archived(
+            contractId,
+            globalKey,
+            _,
+            eventOffset,
+            eventSequentialId,
+          ) =>
+        logger.debug(
+          s"State events update: Archived(contractId=$contractId, globalKey=$globalKey, offset=$eventOffset, eventSequentialId=$eventSequentialId)"
+        )
+      case LedgerEndMarker(eventOffset, eventSequentialId) =>
+        logger.debug(
+          s"Ledger end reached: $eventOffset -> $eventSequentialId"
+        )
+    }
+}
+
+object MutableCacheBackedContractStore {
+  type EventSequentialId = Long
+
+  def apply(
+      contractsReader: LedgerDaoContractsReader,
+      signalNewLedgerHead: Offset => Unit,
+      metrics: Metrics,
+      maxContractsCacheSize: Long = 100000L,
+      maxKeyCacheSize: Long = 100000L,
```

Comment:
Which story in https://digitalasset.atlassian.net/browse/DPP-173 will take care of making these configurable?

Also: we are measuring the number of entries here and not their size in bytes, correct? If we're under time-pressure, I'd be happy to drop the byte-based configuration.

Reply:
It is implemented in the follow-up PR https://github.com/digital-asset/daml/pull/9395